### PR TITLE
Implement more support for per-channel quantization

### PR DESCRIPTION
**Pull request by dzhulgakov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26339 Serialization for per channel qtensor
* #26243 Fix _empty_per_channel_affine_quantized to be less hacky
* #26242 Fold activation permutation inside quantized conv operator
* #26241 Fold weight permutation inside quantized conv operator
* **#26240 Implement more support for per-channel quantization**

In particular adds support for empty/empty_like which is needed for memory layouts to work.

Differential Revision: [D17443220](https://our.internmc.facebook.com/intern/diff/D17443220)

### CODE REVIEW

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -217,14 +217,24 @@ Tensor empty_like(
...
-    // TORCH_INTERNAL_ASSERT(self.qscheme(), at::kPerTensorAffine,
-    //                       "empty_like for quantized Tensor only works for
-    //                        PerTensorAffine scheme right now");
-    return at::_empty_affine_quantized(self.sizes(), options,
-                                       self.q_scale(),
-                                       self.q_zero_point(),
-                                       use_memory_format);
+    auto qscheme = self.qscheme();
+    if (qscheme == kPerTensorAffine) {
+      return at::_empty_affine_quantized(self.sizes(), options,
```
Comment (jerryzh168) [inline aten/src/ATen/native/TensorFactories.cpp]: we have different ordering of arguments here...
maybe we can have:
```at::_empty_affine_quantized(q_scale, q_zero_point, sizes, options, use_memory_format);```
and
```at::_empty_per_channel_affine_quantized_like(q_per_channel_scales, q_per_channel_zero_points, q_per_channel_axis, sizes, options, use_memory_format);```

But should do in a separate PR: @dskhudia could you change the order for `_empty_per_channel_affine_quantized_like`?

Comment (jerryzh168) [reply aten/src/ATen/native/TensorFactories.cpp]: oh `q_per_channel_axis` is added in this PR, then @dzhulgakov could you change the order as well?

File: tools/autograd/gen_python_functions.py [hunk 2]
```diff
@@ -33,7 +33,7 @@
     'copy_sparse_to_sparse_', 'copy_',
     'numpy_T',  # this needs to be an attribute in Python, not a function
     'nonzero(_(out|numpy))?',
-    'set_quantizer_',
+    'set_quantizer_', 'q_per_channel_axis',  # return types not supported yet
```
Comment (jerryzh168) [inline tools/autograd/gen_python_functions.py]: why? we probably need this for serializing PerChannelAffine tensors

Comment (jerryzh168) [reply tools/autograd/gen_python_functions.py]: do you mean quantizer here?

Comment (dzhulgakov) [reply tools/autograd/gen_python_functions.py]: apparently we don't support returning int[] :-\ I can look into fixing it properly

Comment (dzhulgakov) [reply tools/autograd/gen_python_functions.py]: ah, I was just lazy - it was easy enough to patch

File: test/test_quantized_tensor.py [hunk 3]
```diff
@@ -172,10 +174,40 @@ def test_qtensor_permute(self):
...
         # compare permuted + dequantized result with original transposed result
-        self.assertTrue(np.allclose(qr2.dequantize().numpy(), r.numpy().T, atol=2 / scale))
+        self.assertTrue(np.allclose(qr2.dequantize().numpy(), r.numpy().transpose([1, 0, 2, 3]), atol=2 / scale))
         # make permuted result contiguous
         self.assertTrue(torch.equal(qr2.contiguous().int_repr(), qr2.int_repr()))
 
+        # change memory format
+        qlast = qr.contiguous(memory_format=torch.channels_last)
+        self.assertEqual(qr.stride(), list(reversed(sorted(qr.stride()))))
+        self.assertNotEqual(qlast.stride(), list(reversed(sorted(qlast.stride()))))
```
Comment (jerryzh168) [inline test/test_quantized_tensor.py]: what is this for? to check the contiguous operation actually happened?

File: test/test_quantized_tensor.py [hunk 4]
```diff
@@ -172,10 +174,40 @@ def test_qtensor_permute(self):
...
+        self.assertTrue(torch.equal(qr.int_repr(), qlast.int_repr()))
+        self.assertTrue(qr.q_scale() == qlast.q_scale())
+        self.assertTrue(qr.q_zero_point() == qlast.q_zero_point())
+        self.assertTrue(np.array_equal(qlast.dequantize().numpy(), qr.dequantize().numpy()))
```
Comment (jerryzh168) [inline test/test_quantized_tensor.py]: I think we have `self.assertEqual` on pytorch tensor as well. so
```
self.assertEqual(qlast.dequantize(), qr.dequantize())
```
should work

File: test/test_quantized_tensor.py [hunk 5]
```diff
@@ -172,10 +174,40 @@ def test_qtensor_permute(self):
...
-        self.assertTrue(np.allclose(qr2.dequantize().numpy(), r.numpy().T, atol=2 / scale))
+        self.assertTrue(np.allclose(qr2.dequantize().numpy(), r.numpy().transpose([1, 0, 2, 3]), atol=2 / scale))
         # make permuted result contiguous
         self.assertTrue(torch.equal(qr2.contiguous().int_repr(), qr2.int_repr()))
 
+        # change memory format
+        qlast = qr.contiguous(memory_format=torch.channels_last)
+        self.assertEqual(qr.stride(), list(reversed(sorted(qr.stride()))))
+        self.assertNotEqual(qlast.stride(), list(reversed(sorted(qlast.stride()))))
+        self.assertTrue(torch.equal(qr.int_repr(), qlast.int_repr()))
```
Comment (jerryzh168) [inline test/test_quantized_tensor.py]: FYI: int_repr() will do contiguous right now, but I think it should preserve the strides,  will fix in https://github.com/pytorch/pytorch/pull/25429 when I have some time.

**Review by jerryzh168 — APPROVED** [review]
LGTM Thanks

File: aten/src/ATen/native/TensorShape.cpp [hunk 6]
```diff
@@ -347,7 +347,13 @@ Tensor as_strided_tensorimpl(const Tensor& self, IntArrayRef size, IntArrayRef s
 
 Tensor as_strided_qtensorimpl(const Tensor& self, IntArrayRef size, IntArrayRef stride, optional<int64_t> storage_offset_) {
   auto storage_offset = storage_offset_.value_or(self.storage_offset());
-  auto result = detail::make_tensor<QTensorImpl>(Storage(self.storage()), self.type_set(), get_qtensorimpl(self)->quantizer());
+  auto quantizer = get_qtensorimpl(self)->quantizer();
+  TORCH_CHECK(
+      quantizer->qscheme() == QScheme::PER_TENSOR_SYMMETRIC ||
```
Comment (jerryzh168) [inline aten/src/ATen/native/TensorShape.cpp]: we don't have `QScheme::PER_TENSOR_SYMMETRIC` in backend. please remove.

see: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/quantized/Quantizer.h#L121

File: aten/src/ATen/native/TensorFactories.cpp (hunk 1 above)
Comment (jerryzh168) [reply aten/src/ATen/native/TensorFactories.cpp]: looks like this is addressed in the later PRs

Comment (dzhulgakov) [reply aten/src/ATen/native/TensorFactories.cpp]: So the order and name for `_empty_per_channel_affine_quantized_like` (_like!) - was weird as a workaround for some codegen peculiarity - https://github.com/pytorch/pytorch/pull/26243 fixes it

I the order is still not the same - TensorOptions and additional arguments are in the wrong order. It's C++ only as the args are marked as kwargs in python. But we should fix it anyway. Since it's an intrusive change - I'll send a separate PR for it after these ones land

File: test/test_quantized_tensor.py (hunk 3 above)
Comment (dzhulgakov) [reply test/test_quantized_tensor.py]: yeah, just making sure. I can remove it, but it doesn't hurt

File: test/test_quantized_tensor.py (hunk 4 above)
Comment (dzhulgakov) [reply test/test_quantized_tensor.py]: ah, good point! when I tried it I typed assertEquals instead of assertEqual and we don't override that one :-P

File: aten/src/ATen/native/TensorFactories.cpp [hunk 7]
```diff
@@ -218,14 +218,24 @@ Tensor empty_like(
...
+                                         self.q_scale(),
+                                         self.q_zero_point(),
+                                         use_memory_format);
+    } else if (qscheme == kPerChannelAffine) {
```
Comment (raghuramank100) [inline aten/src/ATen/native/TensorFactories.cpp]: Do we also need to support kPerChannelSymmetric and kPerTensorSymmetric here?

File: aten/src/ATen/native/TensorShape.cpp [hunk 8]
```diff
@@ -348,7 +348,12 @@ Tensor as_strided_tensorimpl(const Tensor& self, IntArrayRef size, IntArrayRef s
 
 Tensor as_strided_qtensorimpl(const Tensor& self, IntArrayRef size, IntArrayRef stride, optional<int64_t> storage_offset_) {
   auto storage_offset = storage_offset_.value_or(self.storage_offset());
-  auto result = detail::make_tensor<QTensorImpl>(Storage(self.storage()), self.type_set(), get_qtensorimpl(self)->quantizer());
+  auto quantizer = get_qtensorimpl(self)->quantizer();
+  TORCH_CHECK(
+      quantizer->qscheme() == QScheme::PER_TENSOR_AFFINE,
```
Comment (raghuramank100) [inline aten/src/ATen/native/TensorShape.cpp]: Similar comments as above for other qschemes

File: aten/src/ATen/native/quantized/QTensor.cpp [hunk 9]
```diff
@@ -94,8 +94,10 @@ Tensor q_per_channel_zero_points_quant(const Tensor& self) {
       self.options().dtype(at::kLong));
 }
 
-Quantizer* quantizer(const Tensor& self) {
-  return get_qtensorimpl(self)->quantizer().get();
+IntArrayRef q_per_channel_axis_quant(const Tensor& self) {
+  auto quantizer = get_qtensorimpl(self)->quantizer();
+  TORCH_CHECK(quantizer->qscheme() == kPerChannelAffine);
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/QTensor.cpp]: Support for PerChannelSymmetric

File: aten/src/ATen/native/quantized/QTensor.cpp [hunk 10]
```diff
@@ -183,6 +185,10 @@ Tensor& set_quantizer_(Tensor& self, ConstQuantizerPtr quantizer) {
 }
 
 Tensor quantized_clone(const Tensor& self) {
+  // TODO: add per channel support
+  TORCH_INTERNAL_ASSERT(
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/QTensor.cpp]: Similar comments as above

File: test/test_quantized_tensor.py [hunk 11]
```diff
@@ -77,13 +77,15 @@ def test_per_channel_qtensor_creation(self):
         q = torch._empty_per_channel_affine_quantized_like(scales, zero_points, [numel], [ch_axis], dtype=torch.quint8)
         self.assertEqual(scales, q.q_per_channel_scales())
         self.assertEqual(zero_points, q.q_per_channel_zero_points())
+        self.assertEqual([ch_axis], q.q_per_channel_axis())
```
Comment (raghuramank100) [inline test/test_quantized_tensor.py]: Both per_channel and qtensor creation tests should sweep over symmetric and affine quant.

File: test/test_quantized_tensor.py [hunk 12]
```diff
@@ -153,28 +155,58 @@ def quantize_c(data, scales, zero_points):
...
+        qlast = qr.contiguous(memory_format=torch.channels_last)
+        self.assertEqual(qr.stride(), list(reversed(sorted(qr.stride()))))
+        self.assertNotEqual(qlast.stride(), list(reversed(sorted(qlast.stride()))))
+        self.assertEqual(qr.int_repr(), qlast.int_repr())
+        self.assertEqual(qr.q_scale(), qlast.q_scale())
+        self.assertEqual(qr.q_zero_point(), qlast.q_zero_point())
+        self.assertEqual(qlast.dequantize(), qr.dequantize())
+
+    def test_qtensor_per_channel_permute(self):
+        r = torch.rand(20, 10, 2, 2, dtype=torch.float) * 4 - 2
```
Comment (raghuramank100) [inline test/test_quantized_tensor.py]: Can we allow for permutes that do not change the axis that is quantized?
Would this work?
```python
# x is in nchw
i.e x_q = torch.quantize_linear_per_channel(... axis =[0])
# x_q is also in nchw
x_nhwc =  x_q.permute([0, 2, 3,1])
```

File: torch/_tensor_docs.py [hunk 13]
```diff
@@ -1903,15 +1903,27 @@ def callable(a, b) -> number
...
+elements that matches the corresponding dimensions (from q_per_channel_axis) of
+the tensor.
+""")
+
+add_docstr_all('q_per_channel_axis',
+               r"""
+q_per_channel_zero_points() -> tuple of ints
+
+Given a Tensor quantized by linear (affine) per-channel quantization,
+returns a indices of dimensions on which per-channel quantization is applied.
```
Comment (raghuramank100) [inline torch/_tensor_docs.py]: Do we really support a tuple for axis? Why not restrict axis to be scalar/list of length 1 and clarify this in the API.
Also NIT: the doc string  should say q_per_channel_axis()

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see comments, thanks

File: aten/src/ATen/native/TensorFactories.cpp (hunk 7 above)
Comment (dzhulgakov) [reply aten/src/ATen/native/TensorFactories.cpp]: again, they are not present in the backend

File: test/test_quantized_tensor.py (hunk 12 above)
Comment (dzhulgakov) [reply test/test_quantized_tensor.py]: as I mentioned in another place, we could, but I think we can add it later in a separate diff. And memory format works today already so it should be sufficient

File: torch/_tensor_docs.py (hunk 13 above)
Comment (dzhulgakov) [reply torch/_tensor_docs.py]: How likely are we to have multidim in the future? Having both apis would be confusing

**Review by raghuramank100 — APPROVED** [review]

File: aten/src/ATen/native/quantized/QTensor.cpp [hunk 14]
```diff
@@ -94,8 +94,10 @@ Tensor q_per_channel_zero_points_quant(const Tensor& self) {
       self.options().dtype(at::kLong));
 }
 
-Quantizer* quantizer(const Tensor& self) {
-  return get_qtensorimpl(self)->quantizer().get();
+IntArrayRef q_per_channel_axis_quant(const Tensor& self) {
```
Comment (ezyang) [inline aten/src/ATen/native/quantized/QTensor.cpp]: This is the very first occurrence of an `IntArrayRef` at return site of a native function, and I don't like it. The returned ref is non-owning. So what's its lifetime tied to? This is not documented anywhere. (In fact, the lifetime is tied to the *quantizer*, not the tensor itself; so the semantics don't even match `sizes()` and `strides()`, whose lifetime is tied to the tensor). Furthermore, this would have to be supported properly in the JIT, as all values in the stack must be owning. C.f. implementation of `sizes()`:

```
    Operator(
        "aten::size(Tensor self) -> int[]",
        [](Stack& stack) {
          RECORD_FUNCTION("size", last(stack, 1));
            
          auto t = std::move(pop(stack)).toTensor();
          pack(stack, t.sizes().vec());
          return 0;
        },    
        aliasAnalysisFromSchema()),
```

I suppose, from first principles, the ability to return a primitive array from a function is desirable functionality, I think we should think carefully about how we want to design this. As a first proposal, the return type of this function should be a `std::vector<>`, not an `IntArrayRef`.

cc @gchanan @zdevito
